### PR TITLE
arm: aarch64: Add support for nested exception handlers

### PR DESCRIPTION
**Pull request by carlocaione**
Add support for nested IRQs, enable the interrupts tests but blacklist
the qemu_cortex_a53 platform for unsupported tests.

Tested against the interrupt nested test by @stephanosio at #23636

### CODE REVIEW

File: tests/arch/arm/arm_runtime_nmi/testcase.yaml
Comment (stephanosio) [inline tests/arch/arm/arm_runtime_nmi/testcase.yaml]: This should be handled with `filter: CONFIG_ARMV6_M_ARMV8_M_BASELINE or CONFIG_ARMV7_M_ARMV8_M_MAINLINE` like the rest of (Cortex-M only) `test/arch/arm` tests.

There is already a commit that addresses this in #23636.
https://github.com/zephyrproject-rtos/zephyr/pull/23636/commits/669069118538a091c60267000bcdbafd5fbf13e6

File: samples/drivers/crypto/sample.yaml
Comment (stephanosio) [inline samples/drivers/crypto/sample.yaml]: Why is this necessary? Is there a technical reason why this test should not/cannot run on the `qemu_cortex_a53`?

**Review by stephanosio — CHANGES_REQUESTED** [review]
Since we are storing ELR and SPSR in the exception stack frame, the following should be removed:
https://github.com/zephyrproject-rtos/zephyr/blob/b7f47f7805e5c6dbf92dc624a588f9a6b283258d/arch/arm/core/aarch64/swap_helper.S#L57-L70

https://github.com/zephyrproject-rtos/zephyr/blob/b7f47f7805e5c6dbf92dc624a588f9a6b283258d/arch/arm/core/aarch64/swap_helper.S#L94-L108

https://github.com/zephyrproject-rtos/zephyr/blob/b7f47f7805e5c6dbf92dc624a588f9a6b283258d/include/arch/arm/aarch64/thread.h#L38-L39

**Comment by carlocaione** [comment]
> Since we are storing ELR and SPSR in the exception stack frame, the following should be removed:

I don't think I get your point. We are storing SPSR and ELR in the ESF only if the context switch is taken. If we want the nested IRQs we have to same SPSR and ELR when entering the ISR and restoring those at exit whether we perform the context switch or not.

File: samples/drivers/crypto/sample.yaml
Comment (carlocaione) [reply samples/drivers/crypto/sample.yaml]: nvm, build failure because I forgot `west update`. Pushing v2 without this line.

**Comment by carlocaione** [comment]
The shippable failure is fixed by https://github.com/zephyrproject-rtos/zephyr/pull/23636/commits/94b9be1918307955a6c0a0ec0398128c3f510eec

**Comment by stephanosio** [comment]
> > Ok, I get your point now but SPSR and ELR in the ESF are needed when the task is first scheduled (see `thread.c`).
> 
> I can probably try to move SPSR and ELR from the ESF to the SP and make it restoring by `z_thread_entry_wrapper`.

It might be helpful to define an "ESF" struct and init the SPSR and ELR using it, as AArch32 does:
https://github.com/zephyrproject-rtos/zephyr/pull/23636/commits/09568860621c4a52b207fc3b3ab125a770ae6017#diff-175fb19793403c98b7ba04af9afac387R106-R107

https://github.com/zephyrproject-rtos/zephyr/pull/23636/commits/09568860621c4a52b207fc3b3ab125a770ae6017#diff-175fb19793403c98b7ba04af9afac387R133-R135

https://github.com/zephyrproject-rtos/zephyr/pull/23636/commits/09568860621c4a52b207fc3b3ab125a770ae6017#diff-175fb19793403c98b7ba04af9afac387L131

**Comment by carlocaione** [comment]
> > > > Ok, I get your point now but SPSR and ELR in the ESF are needed when the task is first scheduled (see `thread.c`).
> > > 
> > > 
> > > I can probably try to move SPSR and ELR from the ESF to the SP and make it restoring by `z_thread_entry_wrapper`.
> > 
> > 
> > It might be helpful to define an "ESF" struct and init the SPSR and ELR using it, as AArch32 does:
> > [0956886#diff-175fb19793403c98b7ba04af9afac387R106-R107](https://github.com/zephyrproject-rtos/zephyr/commit/09568860621c4a52b207fc3b3ab125a770ae6017#diff-175fb19793403c98b7ba04af9afac387R106-R107)
> > [0956886#diff-175fb19793403c98b7ba04af9afac387R133-R135](https://github.com/zephyrproject-rtos/zephyr/commit/09568860621c4a52b207fc3b3ab125a770ae6017#diff-175fb19793403c98b7ba04af9afac387R133-R135)
> > [0956886#diff-175fb19793403c98b7ba04af9afac387L131](https://github.com/zephyrproject-rtos/zephyr/commit/09568860621c4a52b207fc3b3ab125a770ae6017#diff-175fb19793403c98b7ba04af9afac387L131)
> > p.s. that is likely to be useful in fault handler implementation later as well.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/0958673ee1ce289f5348eb68950b0a964bacdd8d/include/arch/arm/aarch64/exc.h#L28-L30
> 
> Here you can add SPSR and ELR.

Yeah, this was my idea as well. Preparing a new PR for this change.

**Comment by carlocaione** [comment]
force pushed new PR rebased on top of @stephanosio work and addressed the comments raised in this story.

File: include/arch/arm/aarch64/exc.h
Comment (stephanosio) [inline include/arch/arm/aarch64/exc.h]: I am not sure about this change.

`__esf`/`__basic_sf`, being a "stack frame", is supposed to match the structure of the stack frame saved during exception/interrupt entry (i.e. the structure of data saved in the stack by `z_arm64_enter_exc`). That does not seem to be the case here.

File: include/arch/arm/aarch64/exc.h
Comment (stephanosio) [inline include/arch/arm/aarch64/exc.h]: This is really only an an artifact of `x0` being the first argument to the thread entry function, which points to the actual thread function.

Under normal circumstances, `elr` is the `pc` of the interrupted context, so there need not be another PC. This should really be just `x0` or regs[0].

File: arch/arm/core/aarch64/thread.c
Comment (stephanosio) [inline arch/arm/core/aarch64/thread.c]: s/z_arm64_pendsv/z_arm64_context_switch/

**Review by stephanosio — COMMENTED** [review]
Looks ok overall. There seem to be a few unusual changes.

File: include/arch/arm/aarch64/exc.h
Comment (carlocaione) [reply include/arch/arm/aarch64/exc.h]: this is not the case for AArch64 and ESF is a misnomer for ARM64. This struct is only used by new thread to correctly set the initial stack, nothing else. After the thread is created and running this struct is never actually used.

File: include/arch/arm/aarch64/exc.h
Comment (carlocaione) [reply include/arch/arm/aarch64/exc.h]: I thought PC was a better suited name but I'll change it back (it was already regs[0])

File: arch/arm/core/aarch64/thread.c
Comment (carlocaione) [reply arch/arm/core/aarch64/thread.c]: ACK

File: include/arch/arm/aarch64/exc.h
Comment (carlocaione) [reply include/arch/arm/aarch64/exc.h]: Probably it's better to rename it. ARC for example uses `init_stack_frame` that could be a better choice.

Comment (carlocaione) [reply include/arch/arm/aarch64/exc.h]: the real problem here is that `typedef struct __esf z_arch_esf_t;` and that's clearly wrong. I'll fix that.

File: boards/arm/qemu_cortex_a53/qemu_cortex_a53.yaml
Comment (stephanosio) [inline boards/arm/qemu_cortex_a53/qemu_cortex_a53.yaml]: Adding `nmi` to the ignored tags list should not be necessary.

The Shippable failure will go away once you cherry-pick this commit into this PR (to add the missing filter in `tests/arch/arm/arm_runtime_nmi`): https://github.com/zephyrproject-rtos/zephyr/pull/23636/commits/669069118538a091c60267000bcdbafd5fbf13e6

**Review by stephanosio — COMMENTED** [review]
All arch-related changes look good to me

File: boards/arm/qemu_cortex_a53/qemu_cortex_a53.yaml
Comment (carlocaione) [reply boards/arm/qemu_cortex_a53/qemu_cortex_a53.yaml]: ouch, missed that.

**Review by stephanosio — APPROVED** [review]
LGTM

**Comment by carlocaione** [comment]
Shippable failure seems unrelated

**Comment by stephanosio** [comment]
> Shippable failure seems unrelated

Yes, unrelated. Re-triggered.

**Comment by ioannisg** [comment]
I'd like to get this rebased after we merge the (minor) fixes to the interrupt test

**Comment by carlocaione** [comment]
> I'd like to get this rebased after we merge the (minor) fixes to the interrupt test

Rebased.

File: tests/kernel/interrupt/src/interrupt_util.h
Comment (ioannisg) [inline tests/kernel/interrupt/src/interrupt_util.h]: Is this supposed to work for ARM-only? If so, some comment or even and ifdef CORTEX_A, R would help

File: tests/kernel/interrupt/src/nested_irq.c
Comment (ioannisg) [inline tests/kernel/interrupt/src/nested_irq.c]: This does not look that CORTEX_A is skipped, right?

File: arch/common/sw_isr_common.c
Comment (ioannisg) [inline arch/common/sw_isr_common.c]: Can we take this commit that removes the assert out of this PR? Seems we have not reached an agreement about that, maybe do it on  a separate commit

File: tests/kernel/interrupt/src/interrupt_util.h
Comment (stephanosio) [reply tests/kernel/interrupt/src/interrupt_util.h]: well, `CONFIG_GIC` is defined to be (ARM) Cortex-only:
https://github.com/zephyrproject-rtos/zephyr/blob/2b441732ff3ba31b1c3bf73497b0aab0279d8011/drivers/interrupt_controller/Kconfig.gic#L6-L8

File: tests/kernel/interrupt/src/nested_irq.c
Comment (stephanosio) [reply tests/kernel/interrupt/src/nested_irq.c]: Correct.

Comment (carlocaione) [reply tests/kernel/interrupt/src/nested_irq.c]: Cortex-A is supported with my latest commit. This could break bisectability though.

File: arch/common/sw_isr_common.c
Comment (stephanosio) [reply arch/common/sw_isr_common.c]: I suppose you mean a separate PR; the problem here is that the test will not pass without that commit.

Comment (ioannisg) [reply arch/common/sw_isr_common.c]: OK, I thought that was only a problem in Cortex-R

File: tests/kernel/interrupt/src/nested_irq.c
Comment (stephanosio) [reply tests/kernel/interrupt/src/nested_irq.c]:
```
tests: kernel: interrupt: Enable nested interrupt test for Cortex-A

This commit enables nested interrupt test for the Cortex-A platforms
that use the ARM Generic Interrupt Controller (GIC).

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
```

The intention here is to enable nested interrupt testing for Cortex-A, which it does.

Comment (ioannisg) [reply tests/kernel/interrupt/src/nested_irq.c]: I mean, the NOTE above seems wrong then.

File: arch/common/sw_isr_common.c
Comment (stephanosio) [reply arch/common/sw_isr_common.c]: This is really a problem with GIC (and all the platforms that use GIC; i.e. Cortex-A and Cortex-R), as described here:
https://github.com/zephyrproject-rtos/zephyr/pull/23518#issuecomment-601072336

File: tests/kernel/interrupt/src/nested_irq.c
Comment (stephanosio) [reply tests/kernel/interrupt/src/nested_irq.c]: > I mean, the NOTE above seems wrong then.

You're right, it seems I forgot to update the comment. I will fix that.

File: arch/common/sw_isr_common.c
Comment (ioannisg) [reply arch/common/sw_isr_common.c]: So I want either to reach the agreement among all ARCH maintainers to remove this assert, or remove it conditionally for GIC, or even ARM && !CORTEX_M

File: tests/kernel/interrupt/src/nested_irq.c
Comment (stephanosio) [reply tests/kernel/interrupt/src/nested_irq.c]: Fixed

File: arch/common/sw_isr_common.c
Comment (stephanosio) [reply arch/common/sw_isr_common.c]: I still have doubts about the effectiveness of this assertion as [I already explained earlier](https://github.com/zephyrproject-rtos/zephyr/pull/23518#issuecomment-601077127), but I will make it `#ifndef CONFIG_GIC` for now so that we are not blocked by it.

Comment (stephanosio) [reply arch/common/sw_isr_common.c]: Fixed

Comment (stephanosio) [reply arch/common/sw_isr_common.c]: https://github.com/zephyrproject-rtos/zephyr/pull/23518#issuecomment-606632953

**Review by ioannisg — APPROVED** [review]

File: arch/common/sw_isr_common.c
Comment (ioannisg) [inline arch/common/sw_isr_common.c]: @andrewboie FYI until we have a generic solution with this ASSERT, I've suggested to conditionally de-activate it for GIC

Comment (stephanosio) [reply arch/common/sw_isr_common.c]: @andrewboie Hopefully, we can seek a long-term solution here: https://github.com/zephyrproject-rtos/zephyr/pull/23518#issuecomment-606632953

Comment (andrewboie) [reply arch/common/sw_isr_common.c]: no objections just please file a GH issue

**Pull request merged**
